README
NOTE: This is a version of the original xbindkeys_config source which has been
modified to work with GTK2 and the latest versions of GCC and GNU ld (which are
respectively 10.2.0 and 2.36.1 as of the time of this writing).  It has been
tested under Arch Linux, where it managed to build and run successfully.

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/


this is a config program for xbindkeys
it can create and modify the file ~/.xbindkeysrc, stop and restart xbindkeys.



-- Required

  XFree86

to compile from source:

	the GTK+ include files and libraries.


-- Instructions for instalation:

You have need xbindkeys program
http://hocwp.free.fr/xbindkeys/xbindkeys.html


xbindkeys_config  instalation:

     make
     su root
     make install

for uninstall
     su root
     make unistall

too use xbindkeys_config type:

xbindkeys_config




-- Contact Info

Laurent Vuibert
lapinator@gmx.net

ChangeLog
xbindkeys_config changelog by Radoslav Georgiev <rgeorgiev583@gmail.com> and Laurent Vuibert <lapinator@gmx.fr>


version:

0.1.4:
	Fix: make the build work with GTK2 and the latest versions of GCC and GNU ld

0.1.3:
	Fix: Crash bug when delete all entry.
	Fix: Build-error with gcc-3.3 (thank to Joerg Jaspert for the patch).
	Add: make unistall
	
0.1.2:  
	Add: Update to xbindkey 1.5.5
	Add: Speed_config (for change config quikly)
	
0.1.1:
	Add: Fix Crash Bugs
	
0.1.0: 
        Add: upgrate to xbindkey 1.4.0
	
0.0.3:
	Add: Undo button
	Add: Command line options
	Add: "View genrated rc" button
	Add: Warning Area
	Add: Drag & drop
	Add: accel button
	
0.0.2:
	Add: read .xbindkeysrc file
	Add: write .xbindkeysrc file
	Add: grab keys:


menu.h
/***************************************************************************
          xbindkeys_config : config programme for xbindkey.
                           -------------------
    begin                : Sat jan 5 14:11:34 CEST 2002
    copyright            : (C) 2002 by Laurent VUIBERT
    email                : lapinator@gmx.fr
***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <gtk/gtk.h>

GtkWidget * xbindkeys_config_menu(GtkAccelGroup *accel_group);


middle.h
/***************************************************************************
          xbindkeys_config : config programme for xbindkey.
                           -------------------
    begin                : Sat jan 5 14:11:34 CEST 2002
    copyright            : (C) 2002 by Laurent VUIBERT
    email                : lapinator@gmx.fr
***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <gtk/gtk.h>

GtkWidget * xbindkeys_config_middle();
GtkWidget * xbindkeys_config_middle_list(gboolean show);

void new_file();
void read_file(char file[]);
void save_file(char file_out[]);
void append_file(char file_in[]);
void view_generated_file();
void undo();
void set_default();
void selectable_clist(gboolean state);
void main_exit(); 

void middle_apply_action(GtkWidget *parent, void *data);


speedc.c
/***************************************************************************
          xbindkeys_config : config programme for xbindkey.
                           -------------------
    begin                : Sat jan 5 14:11:34 CEST 2002
    copyright            : (C) 2002 by Laurent VUIBERT
    email                : lapinator@gmx.fr
***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/


#include <gtk/gtk.h>
#include <stdio.h>
#include <dirent.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>


#include "speedc.h"
#include "middle.h"
#include "version.h"

/* TODO change to dynamic list */
typedef struct list_speedc
{
  char Name[30];
  char File[1024];
} list_speedc;

list_speedc list[30];

int SPEEDC_ROW;

extern GtkWidget *label_warning;

static GtkWidget *Speed_config;

/*****************************************************************************/
void xbindkeys_config_speed_init()
{

  char *p, buf[1024];
  int i;

  DIR *dir;
  struct dirent *dp;


  /* clear the Speed_config */
  gtk_clist_clear (GTK_CLIST(Speed_config));


  /* create $HOME/.xbindkey_config is not exist */
  sprintf(buf,"%s/.xbindkeys_config", getenv ("HOME"));
  dir = opendir(buf);
  if( dir==NULL ) {
    if( errno==ENOENT ){
      i = mkdir( buf, S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
      printf("make file ~/.xbindkeys_config\n");
    }
  } else 
    closedir(dir);
  
  /* create $HOME/.xbindkey_config/speed_c is not exist */
  sprintf(buf,"%s/.xbindkeys_config/speed_c", getenv ("HOME"));
  dir = opendir(buf);
  if( dir==NULL ) {
    if( errno==ENOENT ){
      i = mkdir( buf, S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
      printf("make file ~/.xbindkeys_config/speed_c\n");
    }
  } else 
    closedir(dir);

  /* add the default file */
  p = g_strdup_printf("Default");
  gtk_clist_append(GTK_CLIST(Speed_config),  &p );
 
  sprintf(list[0].Name, "Default");
  sprintf(list[0].File, "%s/.xbindkeysrc", getenv ("HOME"));

  
  /* list all file in $HOME/.xbindkey_config/speed_c */
  if(!(dir = opendir(buf))) { 
    return;
  }
  
  dp = readdir(dir);
  while(dp!=NULL){
    if(dp->d_name[0]!='.'){
      p = g_strdup_printf(dp->d_name);
      gtk_clist_append(GTK_CLIST(Speed_config),  &p );
    }
    dp=readdir(dir);
  }

  closedir(dir);
}



/*****************************************************************************/
void speedc_load(GtkWidget *parent, void *data)
{
  char buf2[1024];
  char *buf[1024];

  if(SPEEDC_ROW==0){
    sprintf(buf2, "%s/.xbindkeysrc", getenv ("HOME"));
    read_file(buf2);
  }
  else {
    gtk_clist_get_text (GTK_CLIST(Speed_config), SPEEDC_ROW, 0, buf); 
    sprintf(buf2, "%s/.xbindkeys_config/speed_c/%s", getenv ("HOME"), buf[0]);
    read_file(buf2);
  }
  return;
}

/*****************************************************************************/
void speedc_apply (GtkWidget *parent, void *data)
{
  middle_apply_action(NULL, NULL);

}
/*****************************************************************************/
void speed_select_row (GtkCList *clist, gint row, gint column,
		       GdkEventButton *event, gpointer user_data)
{
  SPEEDC_ROW=row;
  return;
}

/*****************************************************************************/
void speed_select_row_show(GtkCList *clist, gint row, gint column,
			   GdkEventButton *event, gpointer user_data)
{
  SPEEDC_ROW=row;
  speedc_load(NULL, NULL);
}

/*****************************************************************************/
void speedc_save_ok(GtkWidget *parent, GtkWidget *label)
{
  char buf2[1024];

  sprintf(buf2, "%s/.xbindkeys_config/speed_c/%s", getenv ("HOME"),
	  gtk_entry_get_text(GTK_ENTRY(label)));
  save_file(buf2);
  xbindkeys_config_speed_init();
  return;
}

/*****************************************************************************/
void speedc_save_cancel(GtkWidget *parent, void *data)
{
  printf("cancel\n");
}


/*****************************************************************************/
void speedc_save(GtkWidget *parent, void *data)
{
  GtkWidget *window;
  GtkWidget *vbox;
  GtkWidget *hbox;
  GtkWidget *label;
  GtkWidget *button;


  window = gtk_window_new( GTK_WINDOW_TOPLEVEL );
  gtk_window_set_title(GTK_WINDOW(window),
		       "Xbindkeys Config Version : " XBINDKEYS_CONFIG_VERSION );
  
  vbox = gtk_vbox_new (FALSE, 4);
  gtk_container_add(GTK_CONTAINER(window),vbox);
  
  /* The txt */
  hbox = gtk_hbox_new (FALSE, 4);
  gtk_container_add(GTK_CONTAINER(vbox),hbox);
  
  label = gtk_label_new("Save As :");
  gtk_container_add(GTK_CONTAINER(hbox),label);
  
  label =  gtk_entry_new_with_max_length(16);
  gtk_container_add(GTK_CONTAINER(hbox),label);
  
  /* Button */
  hbox = gtk_hbox_new (FALSE, 4);
  gtk_container_add(GTK_CONTAINER(vbox),hbox);
  
  button = gtk_button_new_with_label("Save");
  gtk_container_add(GTK_CONTAINER(hbox),button);
  gtk_signal_connect(GTK_OBJECT(button), "pressed",
		     (GtkSignalFunc)speedc_save_ok, label);
  gtk_signal_connect_object(GTK_OBJECT(button), "pressed",
			    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(window));
  
  
  button = gtk_button_new_with_label("Cancel");
  gtk_container_add(GTK_CONTAINER(hbox),button);
  gtk_signal_connect(GTK_OBJECT(button), "pressed",
		     (GtkSignalFunc)speedc_save_cancel, NULL);
  gtk_signal_connect_object(GTK_OBJECT(button), "pressed",
			    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(window));
  
  gtk_widget_show_all(window);
  
  return;
}


/*****************************************************************************/
void speedc_del(GtkWidget *parent, void *data)
{
  char buf2[1024];
  char *buf[1024];

  if(SPEEDC_ROW==0){
    gtk_label_set_text (GTK_LABEL(label_warning),
			"\nDefault cannot be deleted\n");
  }
  else {
    gtk_clist_get_text (GTK_CLIST(Speed_config), SPEEDC_ROW, 0, buf); 
    sprintf(buf2, "%s/.xbindkeys_config/speed_c/%s", getenv ("HOME"), buf[0]);
    unlink(buf2);
    xbindkeys_config_speed_init();
  }

  return;
}


/*****************************************************************************/
GtkWidget * xbindkeys_config_speed_config(int show)
{
  GtkWidget *frame;
  GtkWidget *src;
  GtkWidget *vbox;
  GtkWidget *vbox2;
  GtkWidget *hbox;
  GtkWidget *button;

  vbox = gtk_vbox_new (FALSE, 4);
  vbox2 = gtk_vbox_new (FALSE, 4);


  frame =  gtk_frame_new ("Speed Config:");
  gtk_container_add(GTK_CONTAINER(vbox), frame);
  
  src = gtk_scrolled_window_new ( NULL, NULL );
  gtk_container_add(GTK_CONTAINER(frame), vbox2);
  gtk_box_pack_start(GTK_BOX(vbox2), src, TRUE, TRUE, 0);
  
  Speed_config = gtk_clist_new (1);

  if(show)
    gtk_signal_connect(GTK_OBJECT(Speed_config), "select-row",
		       (GtkSignalFunc)speed_select_row_show, NULL);
  else 
    gtk_signal_connect(GTK_OBJECT(Speed_config), "select-row",
		       (GtkSignalFunc)speed_select_row, NULL);

  gtk_clist_set_selection_mode (GTK_CLIST(Speed_config), GTK_SELECTION_BROWSE);
  gtk_clist_set_reorderable(GTK_CLIST(Speed_config), TRUE);
  
  gtk_container_add (GTK_CONTAINER(src), Speed_config);
  

  hbox = gtk_hbox_new (FALSE, 4);
  
  if(show){
    button = gtk_button_new_with_label("Apply");
    gtk_container_add(GTK_CONTAINER(hbox), button);
    gtk_signal_connect(GTK_OBJECT(button), "pressed",
		       (GtkSignalFunc)speedc_apply, NULL);
  }
  else {

    button = gtk_button_new_with_label("Load");
    gtk_container_add(GTK_CONTAINER(hbox), button);
    gtk_signal_connect(GTK_OBJECT(button), "pressed",
		       (GtkSignalFunc)speedc_load, NULL);


    button = gtk_button_new_with_label("Save");
    gtk_container_add(GTK_CONTAINER(hbox), button);
    gtk_signal_connect(GTK_OBJECT(button), "pressed",
		       (GtkSignalFunc)speedc_save, NULL);
    
    
    button = gtk_button_new_with_label("Del");
    gtk_container_add(GTK_CONTAINER(hbox), button);
    gtk_signal_connect(GTK_OBJECT(button), "pressed",
		       (GtkSignalFunc)speedc_del, NULL);
  }

  gtk_box_pack_start(GTK_BOX(vbox2), hbox, FALSE, FALSE, 0);
  xbindkeys_config_speed_init();

  return vbox;
}



xbindkeys_config.c
/***************************************************************************
          xbindkeys_config : config programme for xbindkey.
                           -------------------
    begin                : Sat jan 5 14:11:34 CEST 2002
    copyright            : (C) 2002 by Laurent VUIBERT
    email                : lapinator@gmx.fr
***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <gtk/gtk.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "menu.h"
#include "middle.h"
#include "version.h"
#include "speedc.h"

int main (int argc, char *argv[])
{
   GtkWidget *window;
   GtkWidget *menu;
   GtkWidget *middle;
   GtkWidget *vbox;
   GtkWidget *hbox;
   GtkAccelGroup *accel_group;
   char buf[1024];
   int i;
   gboolean show=FALSE;

   sprintf(buf, "%s/.xbindkeysrc", getenv ("HOME"));

   for(i=1; i<argc; i++){
     if(!strcmp(argv[i],"-h") | !strcmp(argv[i],"--help")){
       printf("XBKC: XBindKeys_Config " XBINDKEYS_CONFIG_VERSION " by Laurent Vuibert/n" \
	      "usage: [--version|-v][--file|-f file][--help|-h][--show|-s]\n" \
	      "\n" \
	      "  -h, --help     This Help\n" \
	      "  -f, --file     Use an alternative rc file\n" \
	      "  -s, --show     show only the rc file\n\n");
       exit(0);
     }
     if(!strcmp(argv[i],"-s") | !strcmp(argv[i],"--show"))
       show=TRUE;

     if(!strcmp(argv[i],"-f") | !strcmp(argv[i],"--file")){
       if(i+1<argc) strcpy(buf,argv[++i]);
       else printf("error: use xbindkey_config %s \"FILE\"\n", argv[i]);
     }
   }
   
   gtk_init( &argc, &argv );   

   if(show){
     window = gtk_window_new( GTK_WINDOW_TOPLEVEL );
     gtk_window_set_title(GTK_WINDOW(window),
			  "Xbindkeys Config  *** Show ***  Version : " XBINDKEYS_CONFIG_VERSION );
     gtk_signal_connect(GTK_OBJECT(window), "delete_event",
			(GtkSignalFunc)main_exit, NULL);
     gtk_signal_connect(GTK_OBJECT(window), "destroy",
			(GtkSignalFunc)main_exit, NULL);
     
     hbox    = gtk_hbox_new(FALSE, 0);
     gtk_container_add(GTK_CONTAINER(window),hbox);

     middle = xbindkeys_config_middle_list(show);
     gtk_container_add(GTK_CONTAINER(hbox),middle);

     middle = xbindkeys_config_speed_config(1);
     gtk_container_add(GTK_CONTAINER(hbox),middle);

     read_file(buf);
     selectable_clist(FALSE);
   }
   else{
     window = gtk_window_new( GTK_WINDOW_TOPLEVEL );
     gtk_window_set_title(GTK_WINDOW(window),
			  "Xbindkeys Config Version : " XBINDKEYS_CONFIG_VERSION );
     gtk_signal_connect(GTK_OBJECT(window), "delete_event",
			(GtkSignalFunc)main_exit, NULL);
     gtk_signal_connect(GTK_OBJECT(window), "destroy",
			(GtkSignalFunc)main_exit, NULL);
   
     vbox    = gtk_vbox_new(FALSE, 0);
     gtk_container_add(GTK_CONTAINER(window),vbox);
     
     accel_group = gtk_accel_group_new();
     menu=xbindkeys_config_menu(accel_group);
     gtk_box_pack_start(GTK_BOX(vbox),menu,FALSE,FALSE,0);
     gtk_window_add_accel_group(GTK_OBJECT(window), accel_group);
     
     middle= xbindkeys_config_middle();
     gtk_box_pack_start(GTK_BOX(vbox),middle,TRUE,TRUE,0);
     read_file(buf);   
   }
   
   gtk_widget_show_all(window);
   gtk_main();
   return 0;
}


menu.c
/***************************************************************************
          xbindkeys_config : config programme for xbindkey.
                           -------------------
    begin                : Sat jan 5 14:11:34 CEST 2002
    copyright            : (C) 2002 by Laurent VUIBERT
    email                : lapinator@gmx.fr
***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <gtk/gtk.h>
#include <stdio.h>
#include <stdlib.h>
#include "middle.h"
#include "menu.h"
#include "version.h"



void menu_open_file (GtkMenuItem *menuitem, gpointer user_data);
void menu_open_default_file (GtkMenuItem *menuitem, gpointer user_data);
gboolean menu_open_file_ok(GtkFileSelection *Sel);
void menu_save_default_file (GtkMenuItem *menuitem, gpointer user_data);
void menu_save_file (GtkMenuItem *menuitem, gpointer user_data);
gboolean menu_save_file_ok(GtkFileSelection *Sel);
void menu_append_file (GtkMenuItem *menuitem, gpointer user_data);
gboolean menu_append_file_ok(GtkFileSelection *Sel);

void menu_about (GtkMenuItem *menuitem, gpointer user_data);
void menu_manual (GtkMenuItem *menuitem, gpointer user_data);
void lockfilewindow();
void unlockfilewindow();

GtkWidget *OpenAs;
GtkWidget *AppendAs;
GtkWidget *SaveAs;
GtkWidget *About;
GtkWidget *Manual;

GtkWidget *button_undo_menu;

GtkWidget * xbindkeys_config_menu(GtkAccelGroup *accel_group)
{
  GtkWidget *Handle;
  GtkWidget *MenuBar;
  GtkItemFactory *item_factory;
  
  static GtkItemFactoryEntry menu_items[] = {
    {"/_File", NULL, 0, 0, "<Branch>"},
    {"/File/tearoff", NULL, 0, 0, "<Tearoff>"},
    {"/File/New file", "<control>N", new_file, 0},
    {"/File/Open default file", "<control>O", menu_open_default_file, 0},
    {"/File/Save to default file", "<control>S", menu_save_default_file, 0},
    {"/File/Nothing", NULL, 0, 0,"<Separator>"},
    {"/File/Open as ...", "<control><shift>O", menu_open_file, 0},
    {"/File/Append file ...", "<control><shift>A", menu_append_file, 0},
    {"/File/Save as ...", "<control><shift>S", menu_save_file, 0},
    {"/File/Nothing", NULL, 0, 0,"<Separator>"},
    {"/File/Exit (no Saving)", "<control>Q", main_exit, 0},

    {"/_Edit", NULL, 0, 0, "<Branch>"},
    {"/Edit/tearoff", NULL, 0, 0, "<Tearoff>"},
    {"/Edit/Undo", "<control>Z", undo, 0},
    {"/Edit/Nothing", NULL, 0, 0,"<Separator>"},
    {"/Edit/View generated rc file", NULL, view_generated_file, 0},
    {"/Edit/Set to default rc file", NULL, set_default, 0},

/*
    {"/_SpeedConfig", NULL, 0, 0, "<Branch>"},
    {"/SpeedConfig/tearoff", NULL, 0, 0, "<Tearoff>"},
    {"/SpeedConfig/Quik Save", NULL, undo, 0},
    {"/Edit/Nothing", NULL, 0, 0,"<Separator>"},
    {"/Edit/View generated rc file", NULL, view_generated_file, 0},
    {"/Edit/Set to default rc file", NULL, set_default, 0}, */


    {"/_Help", NULL, 0, 0, "<LastBranch>"},
    {"/Help/tearoff", NULL, 0, 0, "<Tearoff>"},
    {"/Help/About ...", "<control>A", menu_about, 0, "<Item>"},
    {"/Help/Manual ...", "<control>M", menu_manual, 0}
  };

  Handle   = gtk_handle_box_new();
  item_factory = gtk_item_factory_new (GTK_TYPE_MENU_BAR,
				       "<Main>",
				       accel_group);

  gtk_item_factory_create_items ( item_factory, 21, menu_items, NULL);
  MenuBar = gtk_item_factory_get_widget( item_factory, "<Main>");
  
  button_undo_menu = gtk_item_factory_get_widget( item_factory, "/Edit/Undo");

  OpenAs = gtk_item_factory_get_widget( item_factory, "/File/Open as ...");
  AppendAs = gtk_item_factory_get_widget( item_factory, "/File/Append file ...");
  SaveAs = gtk_item_factory_get_widget( item_factory, "/File/Save as ...");

  About = gtk_item_factory_get_widget( item_factory, "/Help/About ...");
  Manual = gtk_item_factory_get_widget( item_factory, "/Help/Manual ...");

  gtk_container_add(GTK_CONTAINER(Handle), MenuBar );

  return Handle;
}


void menu_open_file (GtkMenuItem *menuitem, gpointer user_data)
{
  GtkFileSelection *FichierDgl;
  FichierDgl=(GtkFileSelection*) gtk_file_selection_new("Open xbindkeysrc file");

  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->ok_button), "clicked",
                    (GtkSignalFunc)menu_open_file_ok, GTK_OBJECT(FichierDgl));

  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->ok_button), "clicked",
			    (GtkSignalFunc)unlockfilewindow, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->ok_button), "clicked",
			    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(FichierDgl));
  
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->cancel_button), "clicked",
  			    (GtkSignalFunc)unlockfilewindow, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->cancel_button), "clicked",
			    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(FichierDgl));

  gtk_signal_connect(GTK_OBJECT(FichierDgl), "delete_event",
		     (GtkSignalFunc)gtk_widget_destroy, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl), "delete_event",
			    (GtkSignalFunc)unlockfilewindow, NULL);

  gtk_signal_connect(GTK_OBJECT(FichierDgl), "destroy",
		     (GtkSignalFunc)gtk_widget_destroy, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl), "destroy",
			    (GtkSignalFunc)unlockfilewindow, NULL);

  gtk_widget_show(GTK_WIDGET(FichierDgl));

  lockfilewindow();
}

gboolean menu_open_file_ok(GtkFileSelection *Sel)
{
  gchar *nom;
  nom=gtk_file_selection_get_filename(Sel);
  read_file(nom);
  return TRUE;
}

void menu_open_default_file (GtkMenuItem *menuitem, gpointer user_data)
{
  char buf[1024];
  sprintf(buf, "%s/.xbindkeysrc", getenv ("HOME"));
  read_file(buf);
}

void menu_save_default_file (GtkMenuItem *menuitem, gpointer user_data)
{
  char buf[1024];
  sprintf(buf, "%s/.xbindkeysrc", getenv ("HOME"));
  save_file(buf);
}

void menu_save_file (GtkMenuItem *menuitem, gpointer user_data)
{
  GtkFileSelection *FichierDgl;
  FichierDgl=(GtkFileSelection*) gtk_file_selection_new("Save xbindkeysrc file");

  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->ok_button), "clicked",
                    (GtkSignalFunc)menu_save_file_ok, GTK_OBJECT(FichierDgl));
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->ok_button), "clicked",
                    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(FichierDgl));

  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->cancel_button), "clicked",
                    (GtkSignalFunc)unlockfilewindow, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->cancel_button), "clicked",
                    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(FichierDgl));

  gtk_signal_connect(GTK_OBJECT(FichierDgl), "delete_event",
		     (GtkSignalFunc)gtk_widget_destroy, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl), "delete_event",
			    (GtkSignalFunc)unlockfilewindow, NULL);
  gtk_signal_connect(GTK_OBJECT(FichierDgl), "destroy",
		     (GtkSignalFunc)gtk_widget_destroy, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl), "destroy",
			    (GtkSignalFunc)unlockfilewindow, NULL);

  gtk_widget_show(GTK_WIDGET(FichierDgl));
  lockfilewindow();
}

gboolean menu_save_file_ok(GtkFileSelection *Sel)
{
  gchar *nom;
  nom=gtk_file_selection_get_filename(Sel);
  save_file(nom);
  return TRUE;
}

void menu_append_file (GtkMenuItem *menuitem, gpointer user_data)
{
  GtkFileSelection *FichierDgl;
  FichierDgl=(GtkFileSelection*) gtk_file_selection_new("Append xbindkeysrc file");

  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->ok_button), "clicked",
                    (GtkSignalFunc)menu_append_file_ok, GTK_OBJECT(FichierDgl));
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->ok_button), "clicked",
                    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(FichierDgl));

  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->cancel_button), "clicked",
                    (GtkSignalFunc)unlockfilewindow, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl->cancel_button), "clicked",
                    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(FichierDgl));

  gtk_signal_connect(GTK_OBJECT(FichierDgl), "delete_event",
		     (GtkSignalFunc)gtk_widget_destroy, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl), "delete_event",
			    (GtkSignalFunc)unlockfilewindow, GTK_OBJECT(FichierDgl));
  
  gtk_signal_connect(GTK_OBJECT(FichierDgl), "destroy",
		     (GtkSignalFunc)gtk_widget_destroy, NULL);
  gtk_signal_connect_object(GTK_OBJECT(FichierDgl), "destroy",
			    (GtkSignalFunc)unlockfilewindow, GTK_OBJECT(FichierDgl));
 

  gtk_widget_show(GTK_WIDGET(FichierDgl));
  lockfilewindow();
}

gboolean menu_append_file_ok(GtkFileSelection *Sel)
{
  gchar *nom;
  nom=gtk_file_selection_get_filename(Sel);
  append_file(nom);
  return TRUE;
}

void unlockabout()
{
  gtk_widget_set_sensitive (GTK_WIDGET(About), TRUE);
}

void unlockmanual()
{
  gtk_widget_set_sensitive (GTK_WIDGET(Manual), TRUE);
}


void menu_about (GtkMenuItem *menuitem, gpointer user_data)
{
  GtkWidget *window;
  GtkWidget *text;

  window = gtk_window_new( GTK_WINDOW_TOPLEVEL );
  gtk_window_set_title(GTK_WINDOW(window),
		       "About XBindKeys Config");

  text = gtk_label_new("\n    XBindKeys_Config    \n\n"\
		       "ver : " XBINDKEYS_CONFIG_VERSION "\n\n"\
		       "by Laurent VUIBERT\n"\
		       " lapinator@gmx.fr\n\n");

  gtk_signal_connect(GTK_OBJECT(window), "delete_event",
		     (GtkSignalFunc)unlockabout, NULL);
  gtk_signal_connect_object(GTK_OBJECT(window), "delete_event",
			    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(window));
  gtk_signal_connect(GTK_OBJECT(window), "destroy",
		     (GtkSignalFunc)unlockabout, NULL);
  gtk_signal_connect_object(GTK_OBJECT(window), "delete_event",
			    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(window));

  gtk_container_add(GTK_CONTAINER(window),text);
  gtk_widget_set_sensitive (GTK_WIDGET(About), FALSE);
  gtk_widget_show_all(window);
}

#define MANUAL_TEXT "To edit action, first append item and select them.\n\nTo get key use button \"Get Key\", don't write \nthe key (only for expert).\n\n\n                        FIX ME  (HELP ME)\n\n\n"

void menu_manual (GtkMenuItem *menuitem, gpointer user_data)
{
  GtkWidget *window;
  GtkWidget *text;
  GtkWidget *vbox;
  
  window = gtk_window_new( GTK_WINDOW_TOPLEVEL );
  gtk_window_set_title(GTK_WINDOW(window),
		       "Manual of xbindkeys Config");
  
  vbox = gtk_vbox_new (FALSE,0);
  
  text = gtk_label_new("\nManual\n");
  gtk_box_pack_start(GTK_BOX(vbox), text, FALSE, FALSE, 0);
  
  text = gtk_text_new(NULL,NULL);
  
  
  gtk_text_insert (text, NULL, NULL, NULL,
		   MANUAL_TEXT, sizeof(MANUAL_TEXT)-1);
  gtk_box_pack_start(GTK_BOX(vbox), text, TRUE, TRUE, 0);
  gtk_widget_set_usize(text,300,250);
  
   gtk_signal_connect(GTK_OBJECT(window), "delete_event",
		     (GtkSignalFunc)unlockmanual, NULL);
  gtk_signal_connect_object(GTK_OBJECT(window), "delete_event",
			    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(window));
  gtk_signal_connect(GTK_OBJECT(window), "destroy",
		     (GtkSignalFunc)unlockmanual, NULL);
  gtk_signal_connect_object(GTK_OBJECT(window), "delete_event",
			    (GtkSignalFunc)gtk_widget_destroy, GTK_OBJECT(window));

 
  /*gtk_label_set_justify (GTK_LABEL(text), GTK_JUSTIFY_LEFT);*/

  gtk_container_add(GTK_CONTAINER(window),vbox);
  gtk_widget_set_sensitive (GTK_WIDGET(Manual), FALSE);
  gtk_widget_show_all(window);
}

void lockfilewindow()
{
  gtk_widget_set_sensitive (GTK_WIDGET(OpenAs), FALSE);
  gtk_widget_set_sensitive (GTK_WIDGET(AppendAs), FALSE);
  gtk_widget_set_sensitive (GTK_WIDGET(SaveAs), FALSE);
}

void unlockfilewindow()
{
  gtk_widget_set_sensitive (GTK_WIDGET(OpenAs), TRUE);
  gtk_widget_set_sensitive (GTK_WIDGET(AppendAs), TRUE);
  gtk_widget_set_sensitive (GTK_WIDGET(SaveAs), TRUE);
}
